### DISCUSSION TITLE
Different stream URL for each stream per user

### DISCUSSION THREAD

**Discussion by Byorun** [bluenviron/mediamtx#2877] [Questions]
### Question

Is it possible to have different streaming URL for each stream per user?
> First stream user 1: https://localhost/stream/user1/(something random here)
Second stream user 1: https://localhost/stream/user1/(something other random here)
...
https://localhost/stream/user2/stream_AAXX
https://localhost/stream/user2/stream_XXBB
...

something like that (if possible without extra delay due to restreaming)

**Comment by aler9** [comment]
Hello, in my opinion the best way to achieve what you want is by appending query parameters to the same path:

```
http://localhost/stream?user=user1
http://localhost/stream?user=user2
```

In this way, the same stream will be served to multiple users, avoiding delay due to restreaming, but you will also get access to the user name, that can be parsed by using a reverse proxy before the server or by using an [external authentication server](https://github.com/bluenviron/mediamtx?tab=readme-ov-file#authentication).

**Reply by Byorun** [reply]
That is not quite what i was trying to do. What I am trying to emulate is unique url per stream from each user, much like twitch where each stream from a user has unique stream url. 

right now every stream can be accessed via https://BASEURL/USERNAME (using the integrated mediamtx player) it works and the url stays the same no matter how often the streamer restarts the stream. the goal is that i can differentiate between each stream (per user)

**Reply by aler9** [reply]
You can still use query parameters to achieve what you want. Make users publish to

```
https://baseurl/username
```

Then use `runOnReady` to call a webhook that generates some unique stream ID and stores it inside a database:

```yml
pathDefaults:
  runOnReady: curl http://my-custom-service/generate-stream-id-per-user?user=$MTX_PATH
  runOnNotReady: curl http://my-custom-service/revoke-stream-id-per-user?user=$MTX_PATH
```

Then setup external authentication:

```yml
externalAuthenticationURL: http://my-custom-service/auth
```

Then make readers access the stream with

```
https://baseurl/username?streamid=my-stream-id
```

The server will contact `externalAuthenticationURL` and will send it the stream ID:

```
{
  "ip": "ip",
  "user": "user",
  "password": "password",
  "path": "path",
  "protocol": "rtsp|rtmp|hls|webrtc",
  "id": "id",
  "action": "read|publish",
  "query": "streamid=my-stream-id"
}
```

In your custom service, you can then retrieve the stream ID from the database and check whether it is valid, rejecting the read request if it is not.

**Reply by Byorun** [reply]
I will try that if I have some free time

**Comment by Byorun** [comment]
@aler9 is there a cool way to run multiple commands via runOnReady or only standard linux command chains with "&&" and such?

**Reply by mauricioabreu** [reply]
@Byorun I think the best way to accomplish it is by writing a program (python, go, ruby) etc and call it

**Reply by aler9** [reply]
> only standard linux command chains with "&&" and such?

There are already several ways to launch multiple commands at once (linux &&, windows &, custom software), that is why this feature was not implemented natively into the server.

**Comment by Byorun** [comment]
@aler9 sorry for the long delay.

I tried your method with version 1.8.1, but the query string is only append to the index.m3u8 when I use the bundled hls player. This results in an authentication popup for every stream.m3u8 the player tries to load after that

![image](https://github.com/bluenviron/mediamtx/assets/7345029/e7159f77-e7d9-4ba0-abc4-68ee72a926cb)

**Reply by Byorun** [reply]
ah there is already a bug report for this https://github.com/bluenviron/mediamtx/issues/3275

[a comment by an automated account is not shown]
